### ISSUE TITLE
torch.compile doesn't respect `torch.cuda.set_device()`

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#97280]
repro:
```
@torch.compile
def f():
    a = torch.ones(2, device='cuda')
    torch.cuda.set_device(1)
    return a

print(f())
```

In eager that returns a tensor on the `cuda:0` device, while the above repro returns a tensor on `cuda:1`.

@ngimel pointed out that inductor uses the device stored on the node metadata to figure out how to set device guards in the codegen (`node.meta['val'].device)

What's probably happening is `set_device()` is not part of ATen. So when AOTAutograd traces the dynamo graph into an ATen graph, it directly calls `set_device(1)` instead of putting it into the graph. Ideally, we'd want some way of plumbing the info about the `set_device()` call into inductor

cc @ezyang @soumith @msaroufim @wconstab @ngimel

**Comment by SherlockNoMad** [comment]
Similar to the handling of torch.cuda.stream, we need a general solution for this kind of non-tensor operations program. 

I see there are two kinds of operations in pytorch program,
1. Tensor operations, aka what computations to run. 
2. Run-time configurations, aka how computations should be run. 

Aten operators captures the "whats", and we are missing a good representation for the "hows". 
I don't think adding 'set_device', 'set_stream', 'set_seed' as aten ops is a good solution, as it breaks the abstractions between "tensor operator" and "runtime configurations". 

One alternative representation would be attaching the "runtime configuration" as a fields to the fx.nodes, so that you have 
node.meta["device"] = "cuda:0" 
node,meta["stream"] = "stream1"

such representation has its own limitation. We should discuss more about this.

**Comment by SherlockNoMad** [comment]
cc @yanboliang

**Comment by ngimel** [comment]
Regardless of `set_device` calls before `torch.ones`, the fx graph inductor is getting looks like
```
full: f32[2] = torch.ops.aten.full.default([2], 1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda'), pin_memory = False)
```
(device index is not set). During lowering, inductor decodes device with unset index to be the device that is current during lowering process, so if something executed `torch.cuda.set_device(1)`, current device will be 1. That device will be recorded as the node device and will be used in subsequent codegen. 
Inductor itself cannot recover from this situation, it needs full device (stream, autograd state, what else?) info in fx graph it's getting.

**Comment by yanboliang** [comment]
I like @SherlockNoMad's idea, we can put the runtime configuration(e.g, device, stream, etc) on the FX node.meta when dynamo generating the FX graph, and then all down streams can get and consume it. I can go ahead and give a try if no objection.

**Comment by ngimel** [comment]
We'd need to propagate this to aot tracing, I'm not sure how it will look for backward

**Comment by yanboliang** [comment]
@ngimel If we put them on FX node.meta, can AOT use it in both fwd and bwd?

**Comment by ngimel** [comment]
Hopefully yes

**Comment by SherlockNoMad** [comment]
> We'd need to propagate this to aot tracing, I'm not sure how it will look for backward

We will need to rely upon some meta propagation mechanism to populate them for backward nodes... e.g. the stacktrace population for backward graph. 

In contrast, if we were to represent "runtime config" as an op, we will need to implement the corresponding gradient op for it. The gradient of `set_stream(prev_stream, curr_stream)` will be `set_stream(curr_stream, prev_stream)`

**Comment by bdhirsh** [comment]
If we choose the path of "current context attached to the FX nodes", then in the backward, I think that should be pretty do-able in `ProxyTorchDispatchMode` (from make_fx). As we execute each backward op and create a proxy for it, we can take a look at the global context and record info on the node as necessary.

**Comment by jansel** [comment]
A short term fix might be to graph break on `torch.cuda.set_device` by adding it to:
https://github.com/pytorch/pytorch/blob/1a2dcff127eb0ad39b557b7be1aa64f171e5669c/torch/_dynamo/allowed_functions.py#L83-L110
